From 03dc7d0447301aa564e10af79f94249598a408dc Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Wed, 9 Apr 2025 15:09:06 +0200
Subject: [PATCH] usb: dwc2: gadget: avoid being blocked in case on incomplete
 setup

It fixes earlier patch "ddma enqueue new setup packet when disabled",
when plugging together two dwc2 ports: one as a host, one as device.
It fails to enumerate.
- On device side, Many 'dwc2 48300000.usb: new device is high-speed'
  messages are logged
- On host side:
  usb 1-1: new high-speed USB device number 2 using dwc2
  usb 1-1: device descriptor read/64, error -110
  usb 1-1: device not accepting address 2, error -110
  usb 1-1: new high-speed USB device number 3 using dwc2
  usb 1-1: device descriptor read/64, error -110
  ...
  usb usb1-port1: unable to enumerate USB device

What happens with two dwc2 ports plugged together is that, in ddma mode:
- XferCompl interrupt is raised, without DXEPINT_SETUP flag set
- EP0 has been disabled. With earlier patch, all completed requests are
  scratched, leading to above errors at host side.
  Instead, start a delayed work to ensure setup phase will complete, or
  timeout.
- Setup phase done IRQ is raised (so the data is processed, and the
  delayed work is canceled)

Earlier patch "ddma enqueue new setup packet when disabled" addressed
the incomplete setup phase too suddenly. What lacks in DDMA mode, is
the Timeout IRQ (which isn't raised: just the setup phase done can be
raised). So rather add setup timeout detection by using a delayed work.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: If2b0b1e4c2d03b498cb2706ac14879c0efe36532
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/450440
Reviewed-by: Christian BRUEL <christian.bruel@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/usb/dwc2/core.h   |  1 +
 drivers/usb/dwc2/gadget.c | 40 ++++++++++++++++++++++++++++++---------
 2 files changed, 32 insertions(+), 9 deletions(-)

--- a/drivers/usb/dwc2/core.h
+++ b/drivers/usb/dwc2/core.h
@@ -1101,6 +1101,7 @@ struct dwc2_hsotg {
 	struct workqueue_struct *wq_gadget;
 	struct work_struct wf_otg;
 	struct work_struct wf_gadget;
+	struct delayed_work dw_gsetup;
 	struct timer_list wkp_timer;
 	enum dwc2_lx_state lx_state;
 	struct dwc2_gregs_backup gr_backup;
--- a/drivers/usb/dwc2/gadget.c
+++ b/drivers/usb/dwc2/gadget.c
@@ -3037,6 +3037,20 @@ static void dwc2_gadget_handle_nak(struc
 static void kill_all_requests(struct dwc2_hsotg *hsotg,
 			      struct dwc2_hsotg_ep *ep,
 			      int result);
+
+static void dwc2_gadget_setup_timeout(struct work_struct *work)
+{
+	struct dwc2_hsotg *hsotg = container_of(work, struct dwc2_hsotg, dw_gsetup.work);
+	unsigned long flags;
+
+	dev_dbg(hsotg->dev, "%s: DOEPINT0=0x%08x\n",  __func__, dwc2_readl(hsotg, DOEPINT(0)));
+
+	spin_lock_irqsave(&hsotg->lock, flags);
+	kill_all_requests(hsotg, hsotg->eps_out[0], -ESHUTDOWN);
+	dwc2_hsotg_enqueue_setup(hsotg);
+	spin_unlock_irqrestore(&hsotg->lock, flags);
+}
+
 /**
  * dwc2_hsotg_epint - handle an in/out endpoint interrupt
  * @hsotg: The driver state
@@ -3084,16 +3098,20 @@ static void dwc2_hsotg_epint(struct dwc2
 
 		if (!(dwc2_readl(hsotg, epctl_reg) & DXEPCTL_EPENA)) {
 			/*
-			 * More than 3 setup packets have likely been missed.
-			 * This may happen when other packets has been stored
-			 * in the RxFifo on other endpoints accepting packet
-			 * without being enabled.
-			 * As the EP0 got disabled, complete cb will never be
-			 * called. So enforce here the EP0 remains active (re
-			 * enqueue setup) when later stuck packets gets pulled.
+			 * In DDMA mode, only the "SETUP Phase Done" can be asserted.
+			 * Setup timeout won't be asserted through DXEPINT_SETUP flag.
+			 *
+			 * An incomplete setup phase may be hit when packets have been stored
+			 * into the RxFifo on other endpoints accepting packet without proper
+			 * descriptor chain.
+			 * As the EP0 got disabled, and the DXEPINT_XFERCOMPL flag has been
+			 * cleared above, the complete cb will never be called, due to setup
+			 * timeout not detected.
+			 * So detect here a possible setup request timeout (50ms), to kill
+			 * incomplete setup data, and requeue proper setup descriptors.
 			 */
-			kill_all_requests(hsotg, hsotg->eps_out[0], -ESHUTDOWN);
-			dwc2_hsotg_enqueue_setup(hsotg);
+			queue_delayed_work(hsotg->wq_gadget, &hsotg->dw_gsetup,
+					   msecs_to_jiffies(50));
 		}
 	}
 
@@ -3143,6 +3161,9 @@ static void dwc2_hsotg_epint(struct dwc2
 		dev_dbg(hsotg->dev, "%s: Setup/Timeout\n",  __func__);
 
 		if (using_dma(hsotg) && idx == 0) {
+			/* Safety check to cancel dw_gsetup setup timeout */
+			if (hsotg->ep0_state == DWC2_EP0_SETUP)
+				cancel_delayed_work(&hsotg->dw_gsetup);
 			/*
 			 * this is the notification we've received a
 			 * setup packet. In non-DMA mode we'd get this
@@ -5202,6 +5223,7 @@ int dwc2_gadget_init(struct dwc2_hsotg *
 		return -ENOMEM;
 	}
 	INIT_WORK(&hsotg->wf_gadget, dwc2_gadget_remote_wkup_change);
+	INIT_DELAYED_WORK(&hsotg->dw_gsetup, dwc2_gadget_setup_timeout);
 
 	/* setup endpoint information */
 
